### PR TITLE
[Feature] Adding kappa feature for labor (Cooperative Minibatching)

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
Adding dependency between sampled minibatches increases temporal locality of graph and feature accesses. This can be utilized with the GPU cache PR #4341. #4718 also depends on this PR.

image
image

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by mfbalin** [comment]
@frozenbugs Could you take a look? This implements the kappa feature in the Cooperative Minibatching paper. Has the ability to increase GPU cache hit rates.

**Comment by frozenbugs** [comment]
> @frozenbugs Could you take a look? This implements the kappa feature in the Cooperative Minibatching paper. Has the ability to increase GPU cache hit rates, and reduce feature transfer times.

How did you measure the result? what percentage of increment in GPU cache hit rates and reduction of feature transfer?

**Comment by mfbalin** [comment]
> > @frozenbugs Could you take a look? This implements the kappa feature in the Cooperative Minibatching paper. Has the ability to increase GPU cache hit rates, and reduce feature transfer times.
> 
> How did you measure the result? what percentage of increment in GPU cache hit rates and reduction of feature transfer?

The results are in the cooperative minibatching paper. Basically, can reduce feature transfer times by more than 2x. Will include some of the figures from the paper here as well.

File: src/random/continuous_seed.h [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+#ifdef __CUDA_ARCH__
+  __device__ inline float uniform(const uint64_t t) const {
+    curandStatePhilox4_32_10_t rng;
+    // 123123 is just a number with no significance.
+    curand_init(123123, s[0], t, &rng);
+    float rnd;
+    if (s[0] != s[1]) {
+      rnd = c[0] * curand_normal(&rng);
+      curand_init(123123, s[1], t, &rng);
```
Comment (frozenbugs) [inline src/random/continuous_seed.h:69]: Is it important to have 2 magic number the same, if so, create a local variable at least or create a constant.

Comment (mfbalin) [reply src/random/continuous_seed.h:69]: It is, let me refactor that.

**Review by frozenbugs — APPROVED** [review]

**Comment by mfbalin** [comment]
Just ran the batch_dependency (kappa) path with #4718 and seems to just work.

File: src/random/continuous_seed.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+#ifdef __CUDA_ARCH__
+  __device__ inline float uniform(const uint64_t t) const {
+    const uint64_t MAGIC_CONSTANT = 999961;  // Could be any random number.
+    curandStatePhilox4_32_10_t rng;
+    curand_init(MAGIC_CONSTANT, s[0], t, &rng);
+    float rnd;
+    if (s[0] != s[1]) {
+      rnd = c[0] * curand_normal(&rng);
+      curand_init(MAGIC_CONSTANT, s[1], t, &rng);
```
Comment (frozenbugs) [inline src/random/continuous_seed.h:69]:
```suggestion
    const uint64_t kCurandSeed = 999961;  // Could be any random number.
    curandStatePhilox4_32_10_t rng;
    curand_init(kCurandSeed, s[0], t, &rng);
    float rnd;
    if (s[0] != s[1]) {
      rnd = c[0] * curand_normal(&rng);
      curand_init(kCurandSeed, s[1], t, &rng);
```

**Pull request merged**